scrapy_webarchive/utils.py

```
@@ -78,3 +77,23 @@ def add_ftp_credentials(wacz_uri: str, settings: Settings) -> str:
         return urlunparse(updated_uri)
     
     return wacz_uri
+
+
+def hash_stream(hash_type: str, stream: IO):
+    """Hashes the stream with given hash_type hasher"""
+
+    try:
+        hasher = hashlib.new(hash_type)
+    except (ImportError, ValueError):
+        return 0, ""
```

When would `ImportError` happen? If the hash type is not loaded? Some error message / at least a warning would be appropriate.

---

Refactored this piece a little bit. We now check the `hash_type` first, and return early if the issue occurs while also logging a message. In the hashlib there seem to be 2 sets of algorithms that we can use to check against. 

`algorithms_guaranteed` and `algorithms_available`, from the docs:

`hashlib.algorithms_guaranteed`

A set containing the names of the hash algorithms guaranteed to be supported by this module on all platforms. Note that ‘md5’ is in this list despite some upstream vendors offering an odd “FIPS compliant” Python build that excludes it.

`hashlib.algorithms_available`

A set containing the names of the hash algorithms that are available in the running Python interpreter. These names will be recognized when passed to [new()](https://docs.python.org/3/library/hashlib.html#hashlib.new). [algorithms_guaranteed](https://docs.python.org/3/library/hashlib.html#hashlib.algorithms_guaranteed) will always be a subset. The same algorithm may appear multiple times in this set under different names (thanks to OpenSSL).

`algorithms_guaranteed` seemed to be the safest one to use.

---

Thanks for digging a bit into this. I think `algorithms_available` would make sense here, since we're interested in whether the hash type is available in the current Python interpreter running on the current platform, so we just want to know if we can use this algorithm (and it doesn't matter if a different platform would not support this).

---

> "... so we just want to know if we can use this algorithm (and it doesn't matter if a different platform would not support this)."

But it **does** matter if a different platform supports this or not. We cannot assume the platform these spiders are deployed to. People could introduce OS-specific options that would not be supported on a different platform. Using `hashlib.algorithms_guaranteed` would be safe to use for maximum compatibility. I would suggest leaving this as is.

---

Ah, that is a good point: when we generate an archive, we would like the hash to be readable by the user of the archive.
But then we would need to take into account _all_ software that may handle this, not just Python.

Does datapackage / WACZ mention anything about hasing algorithms?

---

In their examples it is mainly `sha256` that is used, but it does not mention anything about it in the specification. `sha256` is also what is hardcoded as a variable in the `WaczFileCreator` class where the `hash_stream` function is used, without any way to override it (except for subclassing the `WaczFileCreator` class. But if we want to make this adjustable through a setting, it would definitely be good to have a solid implementation of the `hash_stream` function.

---

Ah, that makes sense. No objections to this check, then. Perhaps explain why you do this, and what you try to avoid with this check.

---

Decided to raise an exception instead of logging and returning early. Returning size 0 and an empty hash would be misleading because a user would have to lookup the logs in case something went wrong. Seems to be more appropriate to just raise an exception here.